### PR TITLE
ARROW-17880: [Go] Add support for Decimal128 and Decimal256 to CSV writer

### PR DESCRIPTION
**Pull request by mitchdevenport**

### CODE REVIEW

File: go/arrow/csv/writer.go [hunk 1]
```diff
@@ -219,6 +219,38 @@ func (w *Writer) Write(record arrow.Record) error {
...
+			scale := fieldType.Scale
+			arr := col.(*array.Decimal128)
+			for i := 0; i < arr.Len(); i++ {
+				if arr.IsValid(i) {
+					decString := arr.Value(i).BigInt().Text(10)
+					decimalLocation := len(decString) - int(scale)
+					if decimalLocation >= 0 && scale > 0 {
+						decString = decString[:decimalLocation] + "." + decString[decimalLocation:]
+					}
+					recs[i][j] = decString
```
Comment (zeroshade) [inline go/arrow/csv/writer.go:233]: Any particular reason not to follow the setup used for marshalling to json? 

```go
f := (&big.Float{}).SetInt(arr.Value(i).BigInt())
f.Quo(f, big.NewFloat(math.Pow10(int(scale))))
recs[i][j] = f.Text('g', int(precision))
```
The same thing works for decimal256

Comment (mitchdevenport) [reply go/arrow/csv/writer.go]: None other than I failed to think to check what is done for JSON :)

This is much cleaner, thanks!

Comment (zeroshade) [reply go/arrow/csv/writer.go]: No prob! At this rate I should probably just incorporate this to the `decimal128.Num` and `decimal256.Num` types with a `ToString(precision, scale int32) string` method, but that can be a separate change. :smile:

**Review by zeroshade — APPROVED** [review]
LGTM, when all the tests pass I'll merge it in. Thanks much!!

**Comment by zeroshade** [comment]
@mitchdevenport That said, could you create a JIRA card to add decimal128 and decimal256 to the CSV reader side? (or if you feel like it, you can incorporate it into this PR, that's up to you). It would be great to keep the types that can be written and the types that can be read in sync, but it doesn't have to be in this same change. Thanks again!

**Comment by mitchdevenport** [comment]
> @mitchdevenport That said, could you create a JIRA card to add decimal128 and decimal256 to the CSV reader side? (or if you feel like it, you can incorporate it into this PR, that's up to you). It would be great to keep the types that can be written and the types that can be read in sync, but it doesn't have to be in this same change.
> 
> If you don't end up rolling it into this change, I'll try to find the time to add it to the reader after ApacheCon next week.
> 
> Thanks again!

@zeroshade Yep, no problem! Intended to make it a separate change but was unsure if the original JIRA would close on merge or not. I've updated the original ticket to be specific to the writer and have a filed a new ticket to track the reader changes: https://issues.apache.org/jira/browse/ARROW-17899

**Comment by zeroshade** [comment]
Yes, this ticket will close on merge :)

**Pull request merged**